U/sschmidt23/nzdir

sschmidt23 opened the pull request:
Porting the TXPipe implementation of the NZDir algorithm implemented here: https://github.com/LSSTDESC/TXPipe/blob/nz-dir/txpipe/nz_calibration.py
as it seems more appropriate for RAIL rather than TXPipe
@joezuntz does this meet all of the needs currently covered by the TXPipe version?  This works, and works well for representative data, however, the implementation shortcut of checking radii around spec-z objects rather than photometric objects means that any photometric objects that are not covered by the annuli around the N spec-z objects will not be counted/weighted in the N(z) sum, but will still be present in the bin, which can lead to bias in the estimate. 

examples/estimation/NZDir.ipynb notebook demonstrates use of the algorithm with representative and incomplete data, including bias in the high-z bin for incomplete data.
  
Addresses issue #130

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+    config_options.update(zmin=Param(float, 0.0, msg="The minimum redshift of the z grid"),
+                          zmax=Param(float, 3.0, msg="The maximum redshift of the z grid"),
+                          nzbins=Param(int, 301, msg="The number of gridpoints in the z grid"),
+                          usecols=Param(list, default_usecols, msg="columns from sz_date for Neighor calculation"),
```
typo `sz_date`

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+        if self.config.szweightcol in sz_data.keys():  # pragma: no cover
+            szweights = np.array(sz_data[self.config.szweightcol])
+        else:
+            szweights = np.ones(len(sz_data[self.config.usecols[0]]))
```
I got a bit confused here because I thought the u-band was being used for the weights. My fault for being unobservant, but maybe add a comment explaining that it's unit weights?

sschmidt23 replied:
comment added, and I actually changed this to actually raise an error if the key is asked for but the column is not present in the data file

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+
+    def run(self):
+        if self.config.hdf5_groupname:
+            sz_data = self.get_data('input')[self.config.hdf5_groupname]
```
This is my ignorance of RAIL I think, but does the name "input" get replaced with something else somewhere? Possibly this is the new ceci feature that Eric added.

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+        self.szusecols = None
+        self.szweights = None
+        self.sz_mag_data = None
+        CatEstimator.__init__(self, args, comm=comm)
```
Can also use `super()` here

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+        CatEstimator.__init__(self, args, comm=comm)
+
+    def open_model(self, **kwargs):
+        CatEstimator.open_model(self, **kwargs)
```
Can also use `super()` here

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+
+        # assign weight vecs, if present, else set all to 1.0
+        # tested in example notebook, so just put a pragma no cover for if present
+        if self.config.phot_weightcol in test_data.keys():  # pragma: no cover
```
It looks like if you set `phot_weightcol` but it's not present in the data this will silently be ignored. Is that intended?

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+        self.sz_mag_data = self.model['sz_mag_data']
+
+    def run(self):
+        if self.config.hdf5_groupname:
```
This is the bit that I think changes later when Josue's table_io changes come through. (no change needed now, just thinking through this)

sschmidt23 replied:
Yes, once Josue's changes are in place we can modify this to check chunks of photometry matches within the distance annuli around the spec-z objects and then add the histogram values of the chunks when we finalize.  We'll deal with that once Josue's stuff is finalized, and just leave as-is for now.

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+        else:
+            pweight = np.ones(len(test_data[self.config.usecols[0]]))
+        # calculate weights for the test data
+        if self.config.bincol not in test_data.keys():
```
It would be useful for TXPipe to be able to get the bin col from one file and the photometry data from another (because we will want to try various binnings for the same photometry). I'm not sure if that's possible here, maybe by doing something with the data store?  Maybe this is a TODO.

sschmidt23 replied:
Following Alex's suggestion, I actually removed the multi-bin functionality, so the method now expects a single tomographic bin as input.

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+        else:  # pragma: no cover
+            print("found bin IDs")
+            tomobinids = test_data[self.config.bincol]
+            print(tomobinids)
```
Probably want to remove this printout for laterger runs!

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+        self.add_data('model', self.model)
+
+
+class NZDir(CatEstimator):
```
Is this a summarizer rather than an informer?  Not sure if it makes a difference; maybe they end up the same.

sschmidt23 replied:
There's been a bit of back and forth about the terminology that we use, and what exactly a "summarizer" is.  I think the confusion comes up because we have some tomographic bin estimates which operate using the ensemble of individual galaxy PDFs (e.g the "naive stack" and "point estimate histogram" summarizers), but we also have others, like this NZDir tomographic bin estimate, that do not estimate individual PDFs at all.  I *think* we decided that we can, instead, refer to these latter types of algorithms as "N(z) estimators".  Also, when I decided to split up the "training" of running the initial spec-z tree to determine distance neighborhoods as being separate from the ball tree finding of photometric neighbors, the class structure of catInformer and catEstimator just fit better than the Summarizer classes.  Note that I originally wrote this as a Summarizer class, you can see (look at the commit just before the "switch to inform/estimate rather than summarize" one if you are curious), but felt that this would work better for eventual parallelization, as the checking of photometry neighbors around the spec-z objects should be trivial to run in parallel and then stack up at the end.

joezuntz replied:
Makes sense, thanks for explaining!

joezuntz commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+            single_hist = np.histogram(
+                self.szvec,
+                bins=self.zgrid,
+                range=(0, self.config["zmax"]),
```
I don't think `range` does anything if you set `bins` as well?

sschmidt23 commented on rail/estimation/algos/NZDir.py:
```diff
@@ -0,0 +1,182 @@
+    def __init__(self, args, comm=None):
+        """ Constructor:
+        Do Informer specific initialization """
+        CatInformer.__init__(self, args, comm=comm)
```
super() causes an error for "multiple values of comm" due to how comm is used in the superclass/subclass, so I'll leave them as the specific class names for now, we can revisit that as a separate issue.

sschmidt23 commented:
I removed the simultaneous estimation of multiple tomographic bins via the "bincol" column, so now the estimator expects you to make the cuts first and pass in one bin at a time.  I made appropriate changes to the demo notebook to reflect this, and all runs fine.  I also addressed all of Joe's other minor comments, other than the super().__init __, which is still complaining about comm for some reason.  If we feel like tracking that down then we can do it in a separate issue, as most/all of the estimators have the superclass name hardcoded rather than using super(), and so it is larger than just NZDir.

joezuntz commented:
Looks good!
